Copy any CommentTrivia to simple using statements

06needhamt opened the pull request:
Fixes #37678 

I am very new to the Roslyn codebase so please forgive me if this is not the correct way to approach this issue. I have updated the simple using code fix to copy all the CommentTrivia from inside the block to before the simplified using statement. I would like to make it copy the comments to the respective location in the block but can't seem to work out how.

Any feedback would be welcomed.

CyrusNajmabadi commented:
> I am very new to the Roslyn codebase

Welcome!  and thanks for the contribution! :)

CyrusNajmabadi commented on src/Features/CSharp/Portable/UseSimpleUsingStatement/UseSimpleUsingStatementCodeFixProvider.cs:
```diff
@@ -84,11 +84,25 @@ private static IEnumerable<StatementSyntax> Expand(UsingStatementSyntax usingSta
+            result[0] = result[0].WithLeadingTrivia(comments);
+
             return result;
         }
```
We'd definitely need tests for this change.  They can be added in https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/CSharpTest/UseSimpleUsingStatement/UseSimpleUsingStatementTests.cs

CyrusNajmabadi commented on src/Features/CSharp/Portable/UseSimpleUsingStatement/UseSimpleUsingStatementCodeFixProvider.cs:
```diff
@@ -84,11 +84,25 @@ private static IEnumerable<StatementSyntax> Expand(UsingStatementSyntax usingSta
+                    comments.Add(x);
+                    comments.Add(SyntaxFactory.EndOfLine(Environment.NewLine));
+                }
+            });
```
this seems like it would copy *alllllll* the comments inide the using statement (including trivia associated with internal statements).  this would cause duplications.

CyrusNajmabadi replied:
So, given the bug that we're trying to fix, what we want to do instead is take the leading trivia of the *close curly* of the using statement, and we want to add it as leading trivia to the token that follows the using statement.  that way:

```c#
using (var x = y)
{
   // comment
}

return whatever;
```

Becomes:

```c#
using var x = y;

// comment

return whatever;
```

not:

```c#
// comment
using var x = y;

return whatever;
```

CyrusNajmabadi commented:
I can provide more info if you want.  Or you can reach out at gitter.im/dotnet/roslyn, or in the #roslyn channel in discord.gg/csharp

06needhamt commented:
@CyrusNajmabadi Thanks for the feedback I have changed the behaviour and added a test

CyrusNajmabadi commented on src/Features/CSharp/Portable/UseSimpleUsingStatement/UseSimpleUsingStatementCodeFixProvider.cs:
```diff
@@ -84,9 +84,23 @@ private static IEnumerable<StatementSyntax> Expand(UsingStatementSyntax usingSta
             var result = new List<StatementSyntax>();
             Expand(result, usingStatement);
 
+            var blockStatement = usingStatement.ChildNodes().Where(x => x.Kind() == SyntaxKind.Block);
```
so this won't work for a lot of scenarios.  for example:

```c#
using (var x = y)
using (var a = b)
{
     // whatever
}
```

You'll probably need to update `private static void Expand(List<StatementSyntax> result, UsingStatementSyntax usingStatement)` instead.

CyrusNajmabadi replied:
what you want to do in that function is collect the trivia to preserve when handling the block-syntax case.  it will need to return that trivia here.  that trivia will then need to be returned higher up the stack up to:

```c#
                if (statementToUpdate is UsingStatementSyntax usingStatement &&
                    usingStatement.Declaration != null)
                {
                    var updatedStatements = currentBlock.Statements.ReplaceRange(
                        statementToUpdate,
                        Expand(usingStatement));
                    return currentBlock.WithStatements(updatedStatements);
                }
```

At that point, it will need to take the trivia and assign it to the next statement in the block

06needhamt commented:
@CyrusNajmabadi I have updated the fix to work as you described. However the tests I added seem to fail although the expected output matches what I get when running the code fix. Is there something I am doing wrong/missing.

Thank you for you're patience.

CyrusNajmabadi commented:
I'll try to look into this in the next 1-2 days. Ping me again if I don't!

06needhamt commented:
Thanks 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

CyrusNajmabadi commented:
@06needhamt I've made the requisite fixes here: https://github.com/CyrusNajmabadi/roslyn/tree/patch-37678

You can merge them into your branch.  Cheers!

06needhamt commented:
I’ve merged them in. Thanks for your help. I hope to make more contributions in the future.

CyrusNajmabadi commented:
> . I hope to make more contributions in the future.

That would be fantastic!

CyrusNajmabadi commented:
@jinujoseph can a buddy be assigned for this new community contributer?  Thanks!

CyrusNajmabadi commented:
Also, can you update the title?  It is no longer correct?

06needhamt commented:
@CyrusNajmabadi I have updated the title.

JoeRobich commented on src/Features/CSharp/Portable/UseSimpleUsingStatement/UseSimpleUsingStatementCodeFixProvider.cs:
```diff
@@ -47,9 +48,11 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
             var blocks = topmostUsingStatements.Select(u => (BlockSyntax)u.Parent);
 
+            // Process blocks in reverse order so we rewrite from inside-to-outside with nested
+            // usings.d
```
nit
```suggestion
            // usings.
```

CyrusNajmabadi commented:
@JoeRobich can this be merged in now?  Thanks!

CyrusNajmabadi commented:
Thanks @06needhamt !
